### ISSUE TITLE
Subscriptions expire due to renewal payments not being recorded

### ISSUE THREAD

**Issue by mindctrl** [restrictcontentpro/rcp-braintree#5]
It looks like some return data from Braintree has changed since this was implemented. As a result, `rcp_braintree_webhook()` fails to work properly. A number of PHP notices happen:

```
PHP Notice:  Indirect modification of overloaded property Braintree_Subscription::$transactions has no effect in /html/wp-content/plugins/rcp-braintree/rcp-braintree.php on line 588

PHP Warning:  strtotime() expects parameter 1 to be string, object given in /html/wp-content/plugins/rcp-braintree/rcp-braintree.php on line 590

PHP Notice:  Undefined variable: subscription in /html/wp-content/plugins/rcp-braintree/rcp-braintree.php on line 591

PHP Notice:  Trying to get property of non-object in /html/wp-content/plugins/rcp-braintree/rcp-braintree.php on line 591

WordPress database error Column 'subscription' cannot be null for query INSERT INTO `jp_4_rcp_payments` (`subscription`, `date`, `amount`, `user_id`, `payment_type`, `subscription_key`, `transaction_id`, `status`) VALUES (NULL, '1970-01-01 12:00:00', '5.00', 237, 'Credit Card Subscription', 'jx7jnr', '5tsd39yz', 'complete') made by require('wp-blog-header.php'), require_once('wp-load.php'), require_once('/wp-config.php'), require_once('wp-settings.php'), do_action('init'), call_user_func_array, rcp_braintree_init, do_action('rcp_braintree_webhook'), call_user_func_array, rcp_braintree_webhook, RCP_Payments->insert, QM_DB->query
```

PR incoming

**Comment by mindctrl** [comment]
@pippinsplugins PR attached here. Can you give this a check?

Note, I removed the instance of rcp_insert_payment, as it results in duplicate payments when the webhook comes in. Testing Braintree renewals is a pain in the ass, but I think this will work. I'd say we just defer to payment creation in the webhook anyway. Thoughts?

**Comment by mindctrl** [comment]
PR needs to be adjusted to insert the payment when not auto-renewing.

**Comment by pippinsplugins** [comment]
It looks good to me. Do you have a live site set up to confirm?

**Comment by mindctrl** [comment]
@pippinsplugins yes, works great. Updated to insert the single payments.

**Comment by mindctrl** [comment]
Going to update this again shortly to get the transaction ID from single payments. Just noticed it's not saved.

**Comment by pippinsplugins** [comment]
:+1:

**Comment by mindctrl** [comment]
Rabbit hole! Fixed single payments. Now something changed with date formatting in the updated library it seems. Another push coming in a few.

**Comment by pippinsplugins** [comment]
:D

**Comment by pippinsplugins** [comment]
:+1:

**Comment by mindctrl** [comment]
@pippinsplugins ok I think I'm done here. I hope. 

To update you on this thing, the data structure changed in the updated library (which was required due to a bug in the data returned from subscription payments).

Then occasionally I noticed on subscription payments that the date was either missing or in another location. Only happened a couple of times, and was seemingly the only field that way. I never did catch a log of the payload when it happened, and it's not that easy to reproduce it seems. The date is usually there. I only noticed it in the PHP error log. For that, I set a backup date/time if the date is missing. ¯\_(ツ)_/¯

Everything matches in RCP and Braintree.

Screenshot of payments:
![image](https://cloud.githubusercontent.com/assets/1254870/18186328/369dcfca-7070-11e6-9e3d-c53abe610140.png)

Above you see a regular single credit card payment, followed by a recurring, followed by one that had the funky date thing (which now uses the backup if it's missing).

Hopefully all that's not confusing.
